subsys/net/lib/fota_download/src/fota_download.c

```diff
@@ -30,59 +29,78 @@ static int download_client_callback(const struct download_client_evt *event)
 
 	switch (event->id) {
 	case DOWNLOAD_CLIENT_EVT_FRAGMENT: {
-		size_t size;
+		size_t file_size;
 
-		err = download_client_file_size_get(&dfu, &size);
+		err = download_client_file_size_get(&dlc, &file_size);
 		if (err != 0) {
 			LOG_ERR("download_client_file_size_get error %d", err);
 			callback(FOTA_DOWNLOAD_EVT_ERROR);
 			return err;
 		}
-		if (size > PM_MCUBOOT_SECONDARY_SIZE) {
+
+		if (file_size > PM_MCUBOOT_SECONDARY_SIZE) {
 			LOG_ERR("Requested file too big to fit in flash\n");
 			callback(FOTA_DOWNLOAD_EVT_ERROR);
 			return -EFBIG;
 		}
 
-		err = flash_img_buffered_write(&flash_img,
-				(u8_t *)event->fragment.buf,
-				event->fragment.len, false);
+		if (first_fragment) {
+			first_fragment = false;
+			int img_type = dfu_target_img_type(event->fragment.buf,
+							event->fragment.len);
+
+			err = dfu_target_init(img_type);
+			if (err != 0) {
+				LOG_ERR("dfu_target_init error %d", err);
+				return err;
+			}
+
+			offset = dfu_target_offset();
+			LOG_INF("Offset: 0x%x", offset);
+
+			if (offset != 0) {
+				/* Abort current download procedure, and
+				 * schedule new download from offset.
+				 */
+				k_delayed_work_submit(&restart_dlc_work,
+						K_SECONDS(1));
+				LOG_INF("Refuse fragment, restart with offset");
+
+				return -1;
+			}
+		}
+
+		err = dfu_target_write(event->fragment.buf,
+				       event->fragment.len);
 		if (err != 0) {
-			LOG_ERR("flash_img_buffered_write error %d", err);
+			LOG_ERR("dfu_target_write error %d", err);
+			err = download_client_disconnect(&dlc);
 			callback(FOTA_DOWNLOAD_EVT_ERROR);
 			return err;
 		}
-		break;
+	break;
 	}
 
 	case DOWNLOAD_CLIENT_EVT_DONE:
-		/* Write with 0 length to flush the write operation to flash. */
-		err = flash_img_buffered_write(&flash_img,
-				(u8_t *)event->fragment.buf,
-				0, true);
+		err = dfu_target_done();
 		if (err != 0) {
-			LOG_ERR("flash_img_buffered_write error %d", err);
+			LOG_ERR("dfu_target_done error: %d", err);
 			callback(FOTA_DOWNLOAD_EVT_ERROR);
 			return err;
 		}
 
-		err = boot_request_upgrade(BOOT_UPGRADE_TEST);
-		if (err != 0) {
-			LOG_ERR("boot_request_upgrade error %d", err);
-			callback(FOTA_DOWNLOAD_EVT_ERROR);
-			return err;
-		}
-		err = download_client_disconnect(&dfu);
+		err = download_client_disconnect(&dlc);
 		if (err != 0) {
 			LOG_ERR("download_client_disconncet error %d", err);
 			callback(FOTA_DOWNLOAD_EVT_ERROR);
 			return err;
 		}
-		callback(FOTA_DOWNLOAD_EVT_FINISHED);
+		first_fragment = true;
 		break;
 
 	case DOWNLOAD_CLIENT_EVT_ERROR: {
-		download_client_disconnect(&dfu);
+		download_client_disconnect(&dlc);
+		first_fragment = true;
 		LOG_ERR("Download client error");
 		callback(FOTA_DOWNLOAD_EVT_ERROR);
 		return event->error;
```

Comment:
I would just return `0` here.
Returning non-zero will have a different semantic for `DOWNLOAD_CLIENT_EVT_ERROR` in the near future.
See https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1247

Reply:
We will fix that once #1247 is in. For now, returning non-zero will trigger the same behavior now as when #1247 is merged, which is desirable.